src/dmd/backend/inliner.d

```diff
@@ -0,0 +1,556 @@
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     Distributed under the Boost Software License, Version 1.0.
+ *              http://www.boost.org/LICENSE_1_0.txt
+ * Source:      https://github.com/DigitalMars/Compiler/blob/master/dm/src/dmc/inliner.d
+ */
+
+// C++ specific routines
+
+module dmd.backend.inliner;
+
+version (MARS)
+{
+
+import core.stdc.stdio;
+import core.stdc.ctype;
+import core.stdc.string;
+import core.stdc.stdlib;
+
+//import scopeh;
+
+import dmd.backend.cdef;
+import dmd.backend.cc;
+import dmd.backend.el;
+import dmd.backend.global;
+import dmd.backend.oper;
+import dmd.backend.symtab;
+import dmd.backend.ty;
+import dmd.backend.type;
+
+import dmd.backend.barray;
+import dmd.backend.dlist;
+
+nothrow:
+
+extern (C++)
+{
+//void n2_instantiate_memfunc(Symbol* s);
+//void nwc_mustwrite(Symbol*);
+//void queue_func(Symbol*);
+//elem* cpp_destructor(type* tclass, elem* eptr, elem* enelems, int dtorflag);
+
+//extern __gshared Barray!(elem*) cpp_stidtors;
+
+enum DTORmostderived = 4;
+}
+
+enum log = false;
+
+/**********************************
+ * Determine if function can be inline'd.
+ * Params:
+ *      sfunc = function to check
+ * Returns:
+ *      true if sfunc can be inline'd.
+ */
+
+bool inline_possible(Symbol *sfunc)
+{
+    if (log) printf("inline_possible(%s)\n",sfunc.Sident.ptr);
+    auto f = sfunc.Sfunc;
+    auto t = sfunc.Stype;
+    assert(f && tyfunc(t.Tty));
+
+    bool result = false;
+    if (!(config.flags & CFGnoinlines) && /* if inlining is turned on   */
+        f.Fflags & Finline &&
+        /* Cannot inline varargs or unprototyped functions      */
+        (t.Tflags & (TFfixed | TFprototype)) == (TFfixed | TFprototype) &&
+        !(t.Tty & mTYimport)           // do not inline imported functions
+       )
+    {
+        auto b = f.Fstartblock;
+        if (!b)
+            return false;
+        switch (b.BC)
+        {   case BCret:
+                if (tybasic(t.Tnext.Tty) != TYvoid
+                    && !(f.Fflags & (Fctor | Fdtor | Finvariant))
+                   )
+                {   // Message about no return value
+                    // should already have been generated
+                    break;
+                }
+                goto case BCretexp;
+
+            case BCretexp:
+                result = true;
+                break;
+
+            default:
+                break;
+        }
+    }
+    if (log) printf("returns: %d\n",result);
+    return result;
+}
+
+/**************************
+ * Examine all of the function calls in sfunc, and inline-expand
+ * any that can be.
+ * Params:
+ *      sfunc = function to scan
+ */
+
+void inline_do(Symbol *sfunc)
+{
+    if (log) printf("inline_do(%s)\n",prettyident(sfunc));
+    //symbol_debug(sfunc);
+    func_t* f = sfunc.Sfunc;
+    assert(f && tyfunc(sfunc.Stype.Tty));
+    // BUG: flag not set right in dmd
+    if (1 || f.Fflags3 & Fdoinline)  // if any inline functions called
```

Comment:
> (.text._D4core8internal12parseoptions__T12parseOptionsTSQBv2gc6config6ConfigZQBlFNbNiKQBfAyaZb[_D4core8internal12parseoptions__T12parseOptionsTSQBv2gc6config6ConfigZQBlFNbNiKQBfAyaZb]+0x27): undefined reference to `_TMP1631'

I guess thunks should be marked with `noinline` or whatever internal flag dmd backend has.